feat(gatsby-page-utils): extract logic for watching a directory from gatsby-page-creator so can reuse for custom page creation

frinyvonnick opened the pull request:
## Description

It extract logic from `gatsby-plugin-page-creator` to help creating pages:

- [x] ignorePath
- [x] validatePath
- [x] createPath
- [x] watchDirectory

## Related Issues

Closes #13606

KyleAMathews commented:
Can you add to the descriptions some sample code showing how someone would use this in their gatsby-node.js?

pieh commented:
Should `gatsby-plugin-page-creator` make use of those extracted utilities to not duplicate code?

KyleAMathews commented:
Yeah, what @pieh said. That should be part of this PR.

frinyvonnick commented:
Sure I'll add it as soon as I figure out how to extract watching directory functionnality. I have difficulties to know to limit of things to extract. What function signature did you have in mind @KyleAMathews ?

frinyvonnick commented:
Maybe someone from @gatsbyjs/core could answer my question about function signature of the watching directory functionnality ? 😄

freiksenet commented:
@frinyvonnick Sorry for dropping the ball here!

The API could look something like that:

```js
watchDirectory(
  path: String;
  glob: String;
  onNewFile: (newPath) => void;
  onRemovedFile: (removedPath) => void;
): Promise<undefined>
```

It should track which files were already created and which were not.

frinyvonnick commented:
Thank you @freiksenet, I'll go with it 👌!

frinyvonnick commented:
@KyleAMathews I added some documentation and an example in the [README.md](https://github.com/gatsbyjs/gatsby/blob/dbf0b6dbaa70f9684f3e6a6154f5222ca23ce07e/packages/gatbsy-page-utils/README.md). Is this good for you ? Should I add documentation somewhere else ?

@freiksenet I implemented the signature you suggested. Seems good to you ? 

@gatsbyjs/core can't figure out how to make e2e passes ? Could someone help me on this subject ?

wardpeet commented on packages/gatsby-plugin-page-creator/package.json:
```diff
@@ -21,12 +21,10 @@
   "dependencies": {
```
Not 100% sure but I think the reason why e2e tests are failing are because it can't find the `gatsby-page-utils` package? Make sure to add it do the dependencies here.

It's just a hunch, thanks for taking this on!

wardpeet commented:
mmh let me look into this properly

wardpeet commented:
Fingers crossed 🤞.

gatsby-dev didn't copy the plugin as it was a name mismatch between package.json & foldername. Also the watchDirectory export wasn't a default export but I changed all exports to commonjs (require) to be consistent.

frinyvonnick commented:
Yeah it works :tada: thanks for your help @wardpeet! Sorry for the typo :confused:

frinyvonnick commented on packages/gatsby-page-utils/README.md:
```diff
@@ -0,0 +1,96 @@
+    const createdPath = createPath(addedPath)
+    const page = {
+      path: createdPath,
+      component: systemPath.join(pagesDirectory, filePath),
```
This was a mistake it's addedPath here 😄

frinyvonnick commented on packages/gatsby-page-utils/README.md:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    // Filter out anything matching the given ignore patterns and options
+    if (ignorePath(addedPath, ignore)) {
```
I'm not really sure of a valid value. I put `ignorePath(addedPath, '*/connexion')`. Is this good ?

frinyvonnick commented:
Thank you for your review @muescha 👍

wardpeet commented:
> Yeah it works 🎉 thanks for your help @wardpeet! Sorry for the typo 😕

No need to say sorry, it happens to the best of us 😉 

Thanks for doing this!

frinyvonnick commented:
@wardpeet this is ready to be merged, no ? 😄
